pymess/models/common.py

```diff
@@ -18,3 +29,105 @@ def create_from_related_objects(self, *related_objects):
+    backend = models.CharField(verbose_name=_l('backend'), null=True, blank=True, editable=False, max_length=250)
+    error = models.TextField(verbose_name=_l('error'), null=True, blank=True, editable=False)
+    extra_data = JSONField(verbose_name=_l('extra data'), null=True, blank=True, editable=False)
+    extra_sender_data = JSONField(verbose_name=_l('extra sender data'), null=True, blank=True, editable=False)
+    tag = models.SlugField(verbose_name=_l('tag'), null=True, blank=True, editable=False)
+
+    def __init__(self, *args, **kwargs):
+        assert len(self.STATE.choices) > 0, _l('STATE must be set properly')
+        super().__init__(*args, **kwargs)
+
+    @property
+    def failed(self):
+        return self.state == self.STATE.ERROR
+
+    def __str__(self):
+        return self.recipient
+
+    class Meta:
+        abstract = True
+        ordering = ('-created_at',)
+
+
+class PhoneNumberBasedRecipientMixin:
+
+    def clean_recipient(self):
+        self.recipient = normalize_phone_number(force_text(self.recipient))
+
+    def clean_content(self):
+        pass
+
+
+class BaseRelatedObject(SmartModel):
+    content_type = models.ForeignKey(ContentType, verbose_name=_l('content type of the related object'),
+                                     null=False, blank=False, on_delete=models.CASCADE)
+    object_id = models.TextField(verbose_name=_l('ID of the related object'), null=False, blank=False)
+    object_id_int = models.PositiveIntegerField(verbose_name=_l('ID of the related object in int format'), null=True,
+                                                blank=True, db_index=True)
+    content_object = GenericForeignKey('content_type', 'object_id')
+
+    objects = RelatedObjectManager()
+
+    class Meta:
+        abstract = True
+        ordering = ('-created_at',)
+
+
+class BaseAbstractTemplate(SmartModel):
+
+    slug = models.SlugField(verbose_name=_l('slug'), max_length=100, null=False, blank=False, editable=False,
+                            primary_key=True)
+    body = models.TextField(verbose_name=_l('message body'), null=True, blank=False)
+
+    def _update_context_data(self, context_data):
+        return context_data
+
+    def render_body(self, context_data):
+        context_data = self._update_context_data(context_data)
+        return Template(self.get_body()).render(Context(context_data))
+
+    def clean_body(self, context_data=None):
+        try:
+            self.render_body(context_data or {})
+        except (TemplateSyntaxError, TemplateDoesNotExist) as ex:
+            raise ValidationError(_('Error during template body rendering: "{}"').format(ex))
+
+    def get_body(self):
+        return self.body
+
+    def can_send(self, recipient, context_data):
+        return True
+
+    def send(self, backend_sender, recipient, context_data, related_objects=None, tag=None, **kwargs):
```

Comment:
tohle jde proti OOP, v potomcích měníš vstupní parametry metody. Raději udělej not implemented method get_backend_sender..

Reply:
ak myslis metodu classy, ktora by vracala `backend_sender` a ten parameter by teda vypadol, to som uz skusal. mal som problem s krizovymi importami.

Reply:
no a proč neuděláš import v metodě?

Reply:
tusim som aj to skusal, ale uz si nie som isty. pre istotu preverim

Reply:
zafungovalo to s `get_backend_sender` metodou. asi som ta predtym zle pochopil.